### PR TITLE
feat(pie-input): DSW-1648 integrate assistive text component into input

### PR DESCRIPTION
**Pull request by raoufswe**
## Describe your changes (can list changeset entries if preferable)

[Added] - Integrate assistive text component into input

## Author Checklist (complete before requesting a review)
- [x] I have performed a self-review of my code
- [x] If it is a core feature, I have added thorough tests
- [x] If it is a component change, I have reviewed the Storybook preview
- [x] If there are visual test updates, I have reviewed them properly before approving

### CODE REVIEW

File: packages/components/pie-input/test/component/pie-input.spec.ts [hunk 1]
```diff
@@ -547,6 +553,53 @@ test.describe('PieInput - Component tests', () => {
...
+                });
+
+                // Act
+                const assistiveText = page.locator(assistiveTextSelector);
+
+                // Assert
+                expect(assistiveText).toBeVisible();
+            });
+
+            statusTypes.forEach((status) => {
```
Comment (jamieomaguire) [inline packages/components/pie-input/test/component/pie-input.spec.ts:584]: Please can you add a test that checks that when no status is provided, the assistive text is the default variant if it is provided?

File: packages/components/pie-input/src/defs.ts [hunk 2]
```diff
@@ -1,6 +1,6 @@
 export const types = ['text', 'number', 'password', 'url', 'email', 'tel'] as const;
 export const inputModes = ['none', 'text', 'tel', 'url', 'email', 'numeric', 'decimal', 'search'] as const;
-
+export const statusTypes = ['default', 'error', 'success'] as const;
```
[the remark this answers is not in the document]
Comment (raoufswe) [reply packages/components/pie-input/src/defs.ts]: removed the default type.

File: packages/components/pie-input/test/component/pie-input.spec.ts (hunk 1 above)
Comment (raoufswe) [reply packages/components/pie-input/test/component/pie-input.spec.ts:599]: added

File: packages/components/pie-input/test/visual/pie-input.spec.ts [hunk 3]
```diff
@@ -1,15 +1,60 @@
...
+            {
+                props: { propKeyValues },
+                slots: {
+                    component: testComponent.renderedString.trim(),
+                },
+            },
+        );
+    }));
+
+    // Follow up to remove in Jan
```
Comment (dandel10n) [inline packages/components/pie-input/test/visual/pie-input.spec.ts:56]: Feels like this comment is out of date :)

**Review by dandel10n — DISMISSED** [review]

File: apps/pie-storybook/stories/pie-input.stories.ts [hunk 4]
```diff
@@ -218,7 +238,9 @@ const Template = ({
             ?autoFocus="${autoFocus}"
             ?readonly="${readonly}"
             @input="${onInput}"
-            @change="${onChange}">
+            @change="${onChange}"
+            assistiveText="${ifDefined(assistiveText)}"
+            status=${status || nothing}>
```
Comment (xander-marjoram) [inline apps/pie-storybook/stories/pie-input.stories.ts:243]: Nit: Can we keep the events grouped at the end?
```suggestion
            assistiveText="${ifDefined(assistiveText)}"
            status=${status || nothing}
            @change="${onChange}">
```

File: packages/components/pie-input/README.md [hunk 5]
```diff
@@ -87,6 +87,8 @@ import { PieInput } from '@justeattakeaway/pie-input/dist/react';
 | `readonly` | `boolean` | - | When true, the user cannot edit the control. Not the same as disabled. See [MDN](https://developer.mozilla.org/en-US/docs/Web/HTML/Attributes/readonly) for more information. |
 | `type` | `'text'`, `'number'`, `'password'`, `'url'`, `'email'`, `'tel'` | `text` | The type of HTML input to render. |
 | `value` | `string` | `''` | The value of the input (used as a key/value pair in HTML forms with `name`). |
+| `assistiveText` | `string` | `''` | Allow the user to sets assistive text to display below the input element. |
```
Comment (xander-marjoram) [inline packages/components/pie-input/README.md:90]: For other props when we refer to the user we mean the customer (who won't be setting the assistive text), so it's probably better to be a little more vague
```suggestion
| `assistiveText` | `string` | `''` | Allows assistive text to be displayed below the input element. |
```

File: packages/components/pie-input/src/index.ts [hunk 6]
```diff
@@ -149,6 +159,8 @@ export class PieInput extends FormControlMixin(RtlMixin(LitElement)) implements
             readonly,
             type,
             value,
+            assistiveText,
```
Comment (xander-marjoram) [inline packages/components/pie-input/src/index.ts:162]: Nit: the rest of these are alphabetical

File: packages/components/pie-input/src/index.ts [hunk 7]
```diff
@@ -171,6 +183,7 @@ export class PieInput extends FormControlMixin(RtlMixin(LitElement)) implements
                     @change=${this.handleChange}
                     data-test-id="pie-input">
                 <slot name="trailing"></slot>
+                ${assistiveText ? html`<pie-assistive-text variant=${status || nothing} data-test-id="pie-input-assistive-text">${assistiveText}</pie-assistive-text>` : nothing}
```
Comment (kevinrodrigues) [inline packages/components/pie-input/src/index.ts:186]: I think you can make use of `ifDefined` on the variant attribute?

Comment (xander-marjoram) [inline packages/components/pie-input/src/index.ts:186]: I think it's more appropriate to use `ifDefined` for prop values, something like this:
(I haven't tested this, just did it free-hand)
```suggestion
                ${assistiveText ? html`<pie-assistive-text variant=${ifDefined(status)} data-test-id="pie-input-assistive-text">${assistiveText}</pie-assistive-text>` : nothing}
```

File: packages/components/pie-input/test/component/pie-input.spec.ts [hunk 8]
```diff
@@ -547,6 +553,68 @@ test.describe('PieInput - Component tests', () => {
...
+        test.describe('assistiveText', () => {
+            test('should not render the assistive text component if the prop is not provided', async ({ mount, page }) => {
+                // Arrange
+                await mount(PieInput, {});
+
+                // Act
+                const assistiveText = page.locator(assistiveTextSelector);
+
+                // Assert
+                expect((assistiveText)).not.toBeVisible();
```
Comment (kevinrodrigues) [inline packages/components/pie-input/test/component/pie-input.spec.ts:566]: Do you need the extra parenthesis here?

Comment (xander-marjoram) [inline packages/components/pie-input/test/component/pie-input.spec.ts:566]:
```suggestion
                expect(assistiveText).not.toBeVisible();
```

File: packages/components/pie-input/test/component/pie-input.spec.ts [hunk 9]
```diff
@@ -547,6 +553,68 @@ test.describe('PieInput - Component tests', () => {
...
+                });
+
+                // Act
+                const assistiveText = page.locator(assistiveTextSelector);
+
+                // Assert
+                expect(assistiveText).toBeVisible();
+            });
+
+            test('should render the assistive text component with the default variant if "status" is not provided', async ({ mount, page }) => {
```
Comment (kevinrodrigues) [inline packages/components/pie-input/test/component/pie-input.spec.ts:584]: Not a massive issue so can leave as is but i'd be tempted to merge this test with the one above as it is mostly the same apart from the default check. You can render two expects in the single test and just update the description e.g. `should render the assistive text component if the prop is provided with the default variant if "status" is not provided`

Then the next test could be:

```javascript
describe('Assistive text: Status')
   statusTypes.forEach((status) => {
       test(`should render with the ${status} variant`, async ({ mount, page }) => {

File: packages/components/pie-input/test/visual/pie-input.spec.ts [hunk 10]
```diff
@@ -1,15 +1,60 @@
...
+const componentPropsMatrixByStatus: Record<string, WebComponentPropValues[]> = splitCombinationsByPropertyValue(componentPropsMatrix, 'status');
+const componentStatus: string[] = Object.keys(componentPropsMatrixByStatus);
 
-test.describe('PieInput - Visual tests`', () => {
-    test('should display the PieInput component successfully', async ({ page, mount }) => {
-        await mount(PieInput, {
-            props: {} as InputProps,
-        });
+test.beforeEach(async ({ mount }, testInfo) => {
+    testInfo.setTimeout(testInfo.timeout + 40000);
```
Comment (xander-marjoram) [inline packages/components/pie-input/test/visual/pie-input.spec.ts:31]: Was the `+ 40000` a temporary fix for something?

Comment (raoufswe) [reply packages/components/pie-input/test/visual/pie-input.spec.ts:31]: it's a temporary fix that used in8 files in the codebase 🙄 
maybe @JoshuaNg2332 or @siggerzz has any idea?

File: packages/components/pie-input/test/visual/pie-input.spec.ts (hunk 3 above)
Comment (raoufswe) [reply packages/components/pie-input/test/visual/pie-input.spec.ts:56]: Probably something to @JoshuaNg2332 or @siggerzz as it's referenced in several places.

**Review by fernandofranca — APPROVED** [review]

**Review by dandel10n — APPROVED** [review]

**Pull request merged**